### PR TITLE
fix 10_fused_bias_mask_scale_and_add_fp32

### PR DESCRIPTION
**Pull request by wenxin-zhao**

### CODE REVIEW

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 1]
```diff
@@ -1,79 +1,174 @@
...
+#include <bits/stdc++.h>
 #include "cuda_runtime.h"
-// 注意: 本kernel注释见fp16版本的fused bias mask scale and add
+
+// 实现fp32的fused biasadd mask scale and add的融合算子
+// biasadd + mask + scale + elemwise_add四个算子的融合
+// （x + bias） * mask * scale + addend;
+
+constexpr int n = 100000;
+constexpr int bais_size = 10;
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:10]: 1.我建议还是把L9和L10放在main函数里面的开头，这样看的时候方便些
2.bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 2]
```diff
@@ -1,79 +1,174 @@
...
-    T x_i = x[i] + bias[i % bias_size];
-    y[i] = functor.Compute(x_i, i);
-  }
+__global__ void FusedBaisAdd(FUNCTOR functor, T * dx, T * dy, T * d_bias)
+{
+    int gid = blockDim.x * blockIdx.x + threadIdx.x;
+
+    for (int i = gid; i < n; i += gridDim.x * blockDim.x)
+    {
+        T tmp = dx[i] + d_bias[i % bais_size];
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:38]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 3]
```diff
@@ -1,79 +1,174 @@
...
     }
-    int bias_size = 10;
-    float *x = (float*) malloc(sizeof(float) * ele_cnt);
-    float *y = (float*) malloc(sizeof(float) * ele_cnt);
-    float *bias = (float*) malloc(sizeof(float) * bias_size);
-    for (int i = 0; i < ele_cnt; i++)
+
+    float * x = (float *)malloc(sizeof(float) * n);
+    float * y = (float *)malloc(sizeof(float) * n);
+    float * bias = (float *)malloc(sizeof(float) * bais_size);
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:99]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 4]
```diff
@@ -1,79 +1,174 @@
...
+    for (int i = 0; i < n; ++i)
+    {
+        x[i] = (float)(i);
+        y[i] = 0.0f;
+    }
+    for (int i = 0; i < bais_size; ++i)
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:105]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 5]
```diff
@@ -1,79 +1,174 @@
...
+        bias[i] = i;
+
+    float * groudTruth = (float *)malloc(sizeof(float) * n);
+    for (int i = 0; i < n; ++i)
     {
-      x[i] = (float)(i);
+        groudTruth[i] = (x[i] + bias[i % bais_size]) * static_cast<float>(static_cast<bool>(mask_tensor[i]) * scale) + add_val[i];
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:111]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 6]
```diff
@@ -1,79 +1,174 @@
...
-    float *d_add_val;
-    cudaMalloc((void **)&d_mask_tensor, ele_cnt * sizeof(uint8_t));
-    cudaMalloc((void **)&d_add_val, ele_cnt * sizeof(float));
-    cudaMemcpy(d_mask_tensor, mask_tensor, sizeof(float) * ele_cnt, cudaMemcpyHostToDevice);
-    cudaMemcpy(d_add_val, add_val, sizeof(float) * ele_cnt, cudaMemcpyHostToDevice);
+
+    float * dx, * dy, * d_bias;
+    cudaMalloc((void **)&dx, sizeof(float) * n);
+    cudaMalloc((void **)&dy, sizeof(float) * n);
+    cudaMalloc((void **)&d_bias, sizeof(float) * bais_size);
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:117]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 7]
```diff
@@ -1,79 +1,174 @@
...
+    cudaMemcpy(dx, x, sizeof(float) * n, cudaMemcpyHostToDevice);
+    cudaMemcpy(dy, y, sizeof(float) * n, cudaMemcpyHostToDevice);
+    cudaMemcpy(d_bias, bias, sizeof(float) * bais_size, cudaMemcpyHostToDevice);
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:120]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 8]
```diff
@@ -1,79 +1,174 @@
...
+template<typename FUNCTOR, typename T>
+__global__ void FusedBaisAddVecSmem(FUNCTOR functor, T * dx, T * dy, T * d_bias)
+{
+    int gid = blockDim.x * blockIdx.x + threadIdx.x;
+    int tid = threadIdx.x;
+
+    __shared__ T smem[bais_size];
+
+    // 将d_bias放在shared memory上
+    if (tid < bais_size)
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:53]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu [hunk 9]
```diff
@@ -1,79 +1,174 @@
...
+
+    for (int i = gid; i < n / 4; i += gridDim.x * blockDim.x)
+    {
+        float4 a = reinterpret_cast<float4 *>(dx)[i];
+        float4 b;
+
+        b.x = functor(a.x + smem[(i * 4) % bais_size], i * 4);
+        b.y = functor(a.y + smem[(i * 4 + 1) % bais_size], i * 4 + 1);
+        b.z = functor(a.z + smem[(i * 4 + 2) % bais_size], i * 4 + 2);
+        b.w = functor(a.w + smem[(i * 4 + 3) % bais_size], i * 4 + 3);
```
Comment (RussWong) [inline 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu:65]: bais_size=>bias_size

File: 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu (hunk 1 above)
Comment (wenxin-zhao) [reply 10_fused_bias_mask_scale_and_add/10_fused_bias_mask_scale_and_add_fp32.cu]: 已调整

**Pull request merged**